MyThreadPool/MyThreadPool/MyThreadPool.cs

```diff
@@ -0,0 +1,245 @@
+    /// </summary>
+    public void Shutdown()
+    {
+        if (arrayThreads == null)
+        {
+            throw new ArgumentNullException(nameof(arrayThreads));
+        }
+
+        token.Cancel();
+        waitHandle.Set();
+
+        foreach(var thread in arrayThreads)
+        {
+            thread.Join();
+        }
+    }
+
+    private class MyThread
+    {
+        private Thread? thread;
+        private volatile Queue<Action> tasks;
+        private Object locker;
+        private CancellationToken token;
+        private EventWaitHandle waitHandle;
+
+        /// <summary>
+        /// Task-based custom thread constructor
+        /// </summary>
+        public MyThread(Queue<Action> tasks, CancellationToken token, object locker, EventWaitHandle waitHandle)
+        {
+            this.waitHandle = waitHandle;
+            this.tasks = tasks;
+            this.token = token;
+            this.locker = locker;
+            thread = new Thread(EternalCycle);
+            thread.Start();
+        }
+
+        /// <summary>
+        /// Task waiting cycle
+        /// </summary>
+        private void EternalCycle()
+        {
+            Action? task = null;
+            while (!token.IsCancellationRequested)
+            {
+                waitHandle.WaitOne();
+                if (!token.IsCancellationRequested)
+                {
+                    lock (locker)
+                    {
+                        if (tasks.Count != 0)
+                        {
+                            var isCorrect = tasks.TryDequeue(out task);
+                            if (!isCorrect)
+                            {
+                                throw new InvalidOperationException();
+                            }
+                        }
+                        if (tasks.Count == 0)
+                        {
+                            waitHandle.Reset();
+                        }
+                    }
+                    if (task != null)
+                    {
+                        task();
+                    }
+                    task = null;
+                    if (token.IsCancellationRequested)
+                    {
+                        return;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Suspends the current thread
+        /// </summary>
+        public void Join()
+        {
+            if (thread == null)
+            {
+                throw new ArgumentNullException(nameof(thread));
+            }
+
+            thread.Join();
+        }
+    }
+
+    private class MyTask<TResult> : IMyTask<TResult>
+    {
+        private readonly Func<TResult>? supplier;
+        private TResult? result;
+        private Exception? exception;
+        private CancellationTokenSource token;
+        private EventWaitHandle waitHandle = new EventWaitHandle(false, EventResetMode.ManualReset);
```

Comment:
Эквивалентно
```suggestion
        private ManualResetEvent resetEvent = new ManualResetEvent(false);
```
Удобнее использовать и более распространено :)
Для пропуска одного потока, а не всех сразу, есть AutoResetEvent